src/common/AndroidUtils.ts

```diff
@@ -589,18 +588,44 @@ export class AndroidSDKUtils {
         return `'${pathName}'`;
     }
 
-    private static enableHWKeyboardForEmulator(
+    private static updateEmulatorConfig(
         emulatorName: string
     ): Promise<boolean> {
-        return new Promise((resolve, reject) => {
-            try {
-                shell
-                    .echo('hw.keyboard=yes')
-                    .toEnd(`~/.android/avd/${emulatorName}.avd/config.ini`);
+        return new Promise(async (resolve, reject) => {
+            const config = await AndroidSDKUtils.readEmulatorConfig(
+                emulatorName
+            );
+            if (config.size === 0) {
+                // If we cannot edit the AVD config, fail silently.
+                // This will be a degraded experience but should still work.
                 resolve(true);
-            } catch (error) {
-                reject(error);
+                return;
             }
+
+            // Utilize hardware.
+            config.set('hw.keyboard', 'yes');
+            config.set('hw.gpu.mode', 'auto');
+            config.set('hw.gpu.enabled', 'yes');
+
+            // Give emulator the appropriate skin.
+            let skinName = config.get('hw.device.name') || '';
+            if (skinName) {
```

Comment:
I'm still not sure on what `let skinName = config.get('hw.device.name') || '';` does! Doesn't it mean that `skinName` would either be the result of `config.get(...)` or the empty string if the result of `config.get(...)` was nil or undefined?

If that's what it means then we don't need the extra `if (skinName)` check here since we're guaranteed that it won't be nil/undefined.

Reply:
You are correct, that is what the line does.  I did that because it's cleaner to just deal with string instead of (string || undefined). 

We do need the `if (skinName)` check because you are allowed to add empty string (or undefined) as a map value.  Doing so will make the config invalid and the AVD will not boot.

Reply:
oh i see, I didn't know that `if (skinName)` also checks to ensure `skinName.length>0`. In that case this all LGTM 👍

Reply:
TIL that `myString=""` will product a false-y value for `if (myString)` in JS. Had to try it for myself.  In other news, I miss working in other languages. 😛